lib/pacemaker/pcmk_stonith.c

```diff
@@ -0,0 +1,485 @@
+/*
+ * Copyright 2009-2019 the Pacemaker project contributors
+ *
+ * The version control history for this file may have further details.
+ *
+ * This source code is licensed under the GNU General Public License version 2
+ * or later (GPLv2+) WITHOUT ANY WARRANTY.
+ */
+
+#include <crm_internal.h>
+#include <crm/common/mainloop.h>
+#include <crm/common/output.h>
+#include <crm/stonith-ng.h>
+#include <crm/fencing/internal.h>
+#include <glib.h>
+#include <libxml/tree.h>
+#include <pacemaker.h>
+#include <pcmki/pcmki_stonith.h>
+
+static pcmk__supported_format_t formats[] = {
+    PCMK__SUPPORTED_FORMAT_XML,
+    { NULL, NULL, NULL }
+};
+
+static GMainLoop *mainloop = NULL;
```

Comment:
I hadn't thought about needing a mainloop for some requests. Do you know what happens if the caller has their own mainloop defined? Can two peacefully coexist?

Reply:
Yes, it is possible to have nested main loops.  That's how GTK handles stuff like dialog boxes.  There is a very, very brief mention of it in the glib main loop docs.